toolkit/tools/scheduler/schedulerutils/priority.go

```diff
@@ -0,0 +1,240 @@
+			satisfyingNodes = append(satisfyingNodes, satisfyingNodesForImplicitNode...)
+		}
+	}
+
+	// For each real node we found, we want to prioritize it and all of its dependencies at high.
+	for _, realBuildNode := range satisfyingNodes {
+		nodePriorityMap.SetSubgraphPriority(pkgGraph, realBuildNode, HighNodePriority)
+	}
+
+	return
+}
+
+// IsElevatedPriority returns true if the given node is a medium or high priority node.
+func (s NodePriorityMap) IsElevatedPriority(node *pkggraph.PkgNode) bool {
+	pri := s.priorities[node]
+	return pri > LowNodePriority
+}
+
+// IncreaseNodePriority will increase the priority of the given node to the given priority. If the node already has a
+// higher priority, the priority will not be changed.
+func (s NodePriorityMap) IncreaseNodePriority(node *pkggraph.PkgNode, priority int) {
+	// Limit range of priority to MinPriority and MaxPriority
+	if priority < MinPriority {
+		priority = MinPriority
+	} else if priority > MaxPriority {
+		priority = MaxPriority
+	}
+
+	if s.priorities[node] < priority {
+		logger.Log.Tracef("Setting %s to priority %d", node.FriendlyName(), priority)
+		s.priorities[node] = priority
+	} else {
+		logger.Log.Tracef("Leaving %s at existing priority %d", node.FriendlyName(), s.priorities[node])
+	}
+}
+
+func (s NodePriorityMap) GetPriority(node *pkggraph.PkgNode) (priority int) {
+	return s.priorities[node]
+}
+
+// SetSubgraphPriority will increase the priority of the given node and all of its dependencies to the given priority.
+func (s NodePriorityMap) SetSubgraphPriority(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, priority int) {
+	search := traverse.BreadthFirst{}
+
+	// Walk the graph from the node and prioritize all its dependencies
+	search.Walk(pkgGraph, node, func(n graph.Node, d int) (stopSearch bool) {
+		pkgNode := n.(*pkggraph.PkgNode)
+		logger.Log.Warnf("Setting %s to priority %d", pkgNode.FriendlyName(), priority)
+		s.IncreaseNodePriority(pkgNode, priority)
+		return
+	})
+}
+
+// findMatchesForImplicit will try and find real packages that satisfies the implicit node. This is done through four methods:
+// 1. Find the best package node for the implicit node
+// 2. Try to find a local build node that is backed by the exact same RPM as the implicit node
+// 3. Try 1. again with the version information stripped
+// 4. Finally try 2. again, also with  the version information from the .rpm path removed
+func findMatchesForImplicit(implicitNode *pkggraph.PkgNode, pkgGraph *pkggraph.PkgGraph) (matches []*pkggraph.PkgNode, err error) {
+	const (
+		stripVersion = true
+		matchVersion = false
+	)
+
+	matches, err = findBuildNodesByGraph(implicitNode, pkgGraph, matchVersion)
+	if err != nil {
+		err = fmt.Errorf("failed to find best package node for implicit node (%s):\n%w", implicitNode.FriendlyName(), err)
+		return
+	}
+	if len(matches) > 0 {
+		return
+	}
+
+	matches = findBuildNodesByPath(implicitNode.RpmPath, pkgGraph, matchVersion)
```

Comment:
This part confuses me; I thought that an implicit node, by definition, does not map to a specific RPM, rpmpath, etc. 

An implicit node in my mind is /bin/sh, and we don't know which RPM(s) end up providing that file.

Reply:
At the end of the day every dependency does need to be provided by some rpm. The fetcher asks PMC for all the options that provide an implicit dependency and chooses one.

This path is based on that data we received from PMC when we ran the fetcher. It may or may not be correct, we don't know. But it's probably a decent guess.